rollup/rollup_sync_service/rollup_sync_service.go

```diff
@@ -150,13 +159,39 @@ func (s *RollupSyncService) Stop() {
 		return
 	}
 
+	s.stateMu.Lock()
+	defer s.stateMu.Unlock()
+
 	log.Info("Stopping rollup event sync background service")
 
 	if s.cancel != nil {
 		s.cancel()
 	}
 }
 
+// ResetToHeight resets the RollupSyncService to a specific L1 block height
+func (s *RollupSyncService) ResetToHeight(height uint64) {
+	if s == nil {
+		return
+	}
+
+	s.resetMu.Lock()
+	defer s.resetMu.Unlock()
+
+	s.Stop()
```

Comment:
Why do we need to stop and restart?

Instead, we should just project `latestProcessedBlock` with a lock (i.e. lock in `fetchMessages` and `ResetToHeight`.

Then we also don't need to reinitialize the context here.

Reply:
aha, it makes sense. changed in [c588232](https://github.com/scroll-tech/go-ethereum/pull/1044/commits/c588232a753e61f3c8201b922c34d9db14ccb318), currently, there are some read/write ops related `latestProcessedBlock` and `db`, thus added a function level write lock.

Reply:
refactored to use `atomic.Uint64` for `latestProcessedBlock` in [75413e6](https://github.com/scroll-tech/go-ethereum/pull/1044/commits/75413e676b59f6f528436ab3282dbefbf2d47e76).